src/naip_cnn/sampling.py

```diff
@@ -0,0 +1,258 @@
+"""
+This module contains functions for sampling data from local rasters and Earth Engine 
+images.
+"""
+
+from __future__ import annotations
+
+import io
+import math
+
+import ee
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rasterio
+from joblib import Parallel, delayed
+from PIL import Image
+
+
+def _create_point_grid(*, bounds, spacing: float, crs: str) -> gpd.GeoDataFrame:
+    """Create a regularly spaced grid covering bounds.
+
+    Parameters
+    ----------
+    bounds : tuple
+        The rectangular area to cover (xmin, ymin, xmax, ymax).
+    spacing : float
+        The spacing between points in the grid, in the units of the bounds and CRS.
+    crs : str
+        The CRS of the bounds.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        A dataframe of points covering the bounds in the specified projection.
+    """
+    xmin, ymin, xmax, ymax = bounds
+    x = np.arange(xmin + spacing, xmax - spacing, spacing)
+    y = np.arange(ymin + spacing, ymax - spacing, spacing)
+    xx, yy = np.meshgrid(x, y)
+    points = np.vstack((xx.ravel(), yy.ravel())).T
+    return gpd.GeoDataFrame(
+        geometry=gpd.points_from_xy(points[:, 0], points[:, 1]), crs=crs
+    )
+
+
+def extract_raster_at_points(
+    raster: rasterio.io.DatasetReader, *, points: gpd.GeoDataFrame, band: int = 1
+) -> gpd.GeoDataFrame:
+    """Extract pixel values from a raster at a set of points.
+
+    Parameters
+    ----------
+    raster : rasterio.io.DatasetReader
+        An open raster to sample from.
+    points : gpd.GeoDataFrame
+        A collection of point geometries to sample values at.
+    band : int
+        The raster band index to sample from.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        The original points dataframe with a column added for extracted values. Column
+        names are parsed from the raster band descriptions.
+
+    Raises
+    ------
+    ValueError
+        If the raster CRS does not match the points CRS.
+    """
+    if raster.crs != points.crs:
+        msg = f"The raster CRS ({raster.crs}) and point CRS ({points.crs}) must match."
+        raise ValueError(msg)
+
+    data = raster.read(band)
+    band_name = raster.descriptions[band - 1]

```

Comment:
No, I think you're right that that would break. I was vaguely thinking that `raster.descriptions` can't contain `None`, but that's probably not true if band names aren't set... Probably need to add a fallback here, maybe to `b1` or something similar.